File: android/SentinelPass/app/src/main/java/com/sentinelpass/data/VaultState.kt
package com.sentinelpass.data

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.State
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sentinelpass.Entry
import com.sentinelpass.EntrySummary
import com.sentinelpass.PasswordAnalysis
import com.sentinelpass.TotpCode
import com.sentinelpass.VaultBridge
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.util.UUID

/**
 * Manages vault state and operations
 * Single instance accessible via VaultState.current
 */
class VaultState private constructor(private val context: Context) : ViewModel() {

    private val vaultFile: File
        get() = File(context.filesDir, "sentinelpass_vault.db")

    private val prefs: SharedPreferences
        get() = context.getSharedPreferences("sentinelpass", Context.MODE_PRIVATE)

    private var vaultBridge: VaultBridge? = null

    // UI State
    private val _uiState = MutableStateFlow(VaultUiState())
    val uiState: StateFlow<VaultUiState> = _uiState.asStateFlow()

    // Entries list
    private val _entries = MutableStateFlow<List<EntrySummary>>(emptyList())
    val entries: StateFlow<List<EntrySummary>> = _entries.asStateFlow()

    // Auto-lock timer
    private var autoLockJob: kotlinx.coroutines.Job? = null
    private val autoLockDelayMillis = 5 * 60 * 1000L // 5 minutes

    init {
        _uiState.value = _uiState.value.copy(
            hasVault = vaultFile.exists()
        )
    }

    // ==========================================================================
    // Vault Management
    // ==========================================================================

    /**
     * Create a new vault
     */
    fun createVault(masterPassword: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)

            val result = withContext(Dispatchers.IO) {
                val bridge = VaultBridge(context)
                val success = bridge.initVault(vaultFile.absolutePath, masterPassword)
                if (success) {
                    vaultBridge = bridge
                    prefs.edit().putBoolean("vault_created", true).apply()
                }
                success
            }

            _uiState.value = _uiState.value.copy(
                isLoading = false,
                isUnlocked = result,
                hasVault = result
            )

            if (result) {
                loadEntries()
            }
        }
    }

    /**
     * Unlock existing vault
     */
    fun unlockVault(masterPassword: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)

            val result = withContext(Dispatchers.IO) {
                val bridge = VaultBridge(context)
                val success = bridge.initVault(vaultFile.absolutePath, masterPassword)
                if (success) {
                    vaultBridge = bridge
                }
                success
            }

            _uiState.value = _uiState.value.copy(
                isLoading = false,
                isUnlocked = result,
                error = if (!result) "Invalid master password" else null
            )

            if (result) {
                loadEntries()
            }
        }
    }

    /**
     * Unlock with biometric
     */
    fun unlockWithBiometric() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)

            val result = withContext(Dispatchers.IO) {
                vaultBridge?.unlockWithBiometric() ?: false
            }

            _uiState.value = _uiState.value.copy(
                isLoading = false,
                isUnlocked = result,
                error = if (!result) "Biometric unlock failed" else null
            )

            if (result) {
                loadEntries()
            }
        }
    }

    /**
     * Lock the vault
     */
    fun lockVault() {
        viewModelScope.launch {
            withContext(Dispatchers.IO) {
                vaultBridge?.lockVault()
            }
            vaultBridge = null
            _uiState.value = VaultUiState(hasVault = true)
            _entries.value = emptyList()
        }
    }

    // ==========================================================================
    // Entry Management
    // ==========================================================================

    /**
     * Load all entries from vault
     */
    fun loadEntries() {
        viewModelScope.launch {
            val entries = withContext(Dispatchers.IO) {
                vaultBridge?.listEntries() ?: emptyList()
            }
            _entries.value = entries
        }
    }

    /**
     * Get full entry details by ID
     */
    suspend fun getEntry(id: String): Entry? {
        return withContext(Dispatchers.IO) {
            vaultBridge?.getEntry(id)
        }
    }

    /**
     * Add new entry
     */
    fun addEntry(
        title: String,
        username: String,
        password: String,
        url: String = "",
        notes: String = ""
    ) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)

            val result = withContext(Dispatchers.IO) {
                vaultBridge?.addEntry(title, username, password, url, notes)
            }

            _uiState.value = _uiState.value.copy(
                isLoading = false,
                error = if (result == null) "Failed to add entry" else null
            )

            if (result != null) {
                loadEntries()
            }
        }
    }

    /**
     * Update entry (delete + add for now)
     */
    fun updateEntry(
        id: String,
        title: String,
        username: String,
        password: String,
        url: String = "",
        notes: String = ""
    ) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)

            val result = withContext(Dispatchers.IO) {
                // Delete old entry
                vaultBridge?.deleteEntry(id)
                // Add updated entry
                vaultBridge?.addEntry(title, username, password, url, notes)
            }

            _uiState.value = _uiState.value.copy(
                isLoading = false,
                error = if (result == null) "Failed to update entry" else null
            )

            if (result != null) {
                loadEntries()
            }
        }
    }

    /**
     * Delete entry
     */
    fun deleteEntry(id: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)

            val result = withContext(Dispatchers.IO) {
                vaultBridge?.deleteEntry(id) ?: false
            }

            _uiState.value = _uiState.value.copy(
                isLoading = false,
                error = if (!result) "Failed to delete entry" else null
            )

            if (result) {
                loadEntries()
            }
        }
    }

    /**
     * Search entries
     */
    suspend fun searchEntries(query: String): List<EntrySummary> {
        return withContext(Dispatchers.IO) {
            vaultBridge?.searchEntries(query) ?: emptyList()
        }
    }

    // ==========================================================================
    // TOTP
    // ==========================================================================

    /**
     * Generate TOTP code for entry
     */
    suspend fun generateTotp(entryId: String): TotpCode? {
        return withContext(Dispatchers.IO) {
            vaultBridge?.generateTotp(entryId)
        }
    }

    // ==========================================================================
    // Password Generation
    // ==========================================================================

    /**
     * Generate random password
     */
    suspend fun generatePassword(length: Int, includeSymbols: Boolean): String? {
        return withContext(Dispatchers.IO) {
            vaultBridge?.generatePassword(length, includeSymbols)
        }
    }

    /**
     * Check password strength
     */
    suspend fun checkPasswordStrength(password: String): PasswordAnalysis? {
        return withContext(Dispatchers.IO) {
            vaultBridge?.checkStrength(password)
        }
    }

    // ==========================================================================
    // Biometric
    // ==========================================================================

    /**
     * Check if biometric key exists
     */
    suspend fun hasBiometricKey(): Boolean {
        return withContext(Dispatchers.IO) {
            vaultBridge?.hasBiometricKey() ?: false
        }
    }

    /**
     * Remove biometric key
     */
    fun disableBiometric() {
        viewModelScope.launch {
            withContext(Dispatchers.IO) {
                vaultBridge?.removeBiometricKey()
            }
        }
    }

    // ==========================================================================
    // Auto-Lock
    // ==========================================================================

    /**
     * Schedule auto-lock after delay
     */
    fun scheduleAutoLock() {
        autoLockJob?.cancel()
        autoLockJob = viewModelScope.launch {
            delay(autoLockDelayMillis)
            if (_uiState.value.isUnlocked) {
                lockVault()
            }
        }
    }

    /**
     * Check and execute auto-lock if scheduled
     */
    fun checkAutoLock() {
        // Auto-lock is handled by the scheduleAutoLock coroutine
        // This is called when app returns from foreground
    }

    override fun onCleared() {
        super.onCleared()
        vaultBridge?.destroyVault()
    }

    companion object {
        @Volatile
        private var INSTANCE: VaultState? = null

        fun initialize(context: Context) {
            if (INSTANCE == null) {
                INSTANCE = VaultState(context.applicationContext)
            }
        }

        val current: VaultState
            get() = INSTANCE ?: throw IllegalStateException("VaultState not initialized")
    }
}

/**
 * UI State for vault
 */
data class VaultUiState(
    val hasVault: Boolean = false,
    val isUnlocked: Boolean = false,
    val isLoading: Boolean = false,
    val error: String? = null
)

File: ios/SentinelPass/SentinelPass/Models/VaultState.swift
//
//  VaultState.swift
//  SentinelPass
//
//  Manages vault state and communicates with the bridge
//

import Foundation
import SwiftUI
import LocalAuthentication

@MainActor
class VaultState: ObservableObject {
    static let shared = VaultState()

    @Published var isUnlocked: Bool = false
    @Published var hasVault: Bool = false
    @Published var entries: [EntryModel] = []
    @Published var errorMessage: String?
    @Published var isLoading: Bool = false

    private var vaultBridge: VaultBridge?
    private let vaultURL: URL

    private init() {
        // Vault stored in app's documents directory
        let documentsDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        self.vaultURL = documentsDir.appendingPathComponent("sentinelpass_vault.db")
        self.hasVault = FileManager.default.fileExists(atPath: vaultURL.path)
    }

    // MARK: - Vault Management

    func createVault(masterPassword: String) async throws {
        isLoading = true
        defer { isLoading = false }

        let bridge = VaultBridge()
        let success = await bridge.createVault(
            vaultPath: vaultURL.path,
            masterPassword: masterPassword
        )

        guard success else {
            throw VaultError.creationFailed
        }

        self.vaultBridge = bridge
        self.hasVault = true
        self.isUnlocked = true
        await loadEntries()
    }

    func unlockVault(masterPassword: String) async throws {
        isLoading = true
        defer { isLoading = false }

        let bridge = VaultBridge()
        let success = await bridge.unlockVault(
            vaultPath: vaultURL.path,
            masterPassword: masterPassword
        )

        guard success else {
            throw VaultError.invalidPassword
        }

        self.vaultBridge = bridge
        self.isUnlocked = true
        await loadEntries()
    }

    func lockVault() {
        vaultBridge?.lockVault()
        vaultBridge = nil
        isUnlocked = false
        entries.removeAll()
    }

    // MARK: - Entry Management

    func loadEntries() async {
        guard let bridge = vaultBridge else { return }

        let summaries = await bridge.listEntries()
        entries = summaries.map { summary in
            EntryModel(
                id: summary.id,
                title: summary.title,
                username: summary.username,
                favorite: summary.favorite
            )
        }
    }

    func getEntry(id: String) async throws -> EntryModel {
        guard let bridge = vaultBridge else {
            throw VaultError.vaultLocked
        }

        guard let entry = await bridge.getEntry(id: id) else {
            throw VaultError.entryNotFound
        }

        return EntryModel(
            id: entry.id,
            title: entry.title,
            username: entry.username,
            password: entry.password,
            url: entry.url,
            notes: entry.notes,
            favorite: entry.favorite,
            createdAt: entry.createdAt,
            modifiedAt: entry.modifiedAt
        )
    }

    func addEntry(title: String, username: String, password: String, url: String, notes: String) async throws {
        guard let bridge = vaultBridge else {
            throw VaultError.vaultLocked
        }

        guard let entryId = await bridge.addEntry(
            title: title,
            username: username,
            password: password,
            url: url,
            notes: notes
        ) else {
            throw VaultError.addEntryFailed
        }

        await loadEntries()
    }

    func updateEntry(id: String, title: String, username: String, password: String, url: String, notes: String) async throws {
        guard let bridge = vaultBridge else {
            throw VaultError.vaultLocked
        }

        let success = await bridge.updateEntry(
            id: id,
            title: title,
            username: username,
            password: password,
            url: url,
            notes: notes
        )

        guard success else {
            throw VaultError.updateEntryFailed
        }

        await loadEntries()
    }

    func deleteEntry(id: String) async throws {
        guard let bridge = vaultBridge else {
            throw VaultError.vaultLocked
        }

        let success = await bridge.deleteEntry(id: id)
        guard success else {
            throw VaultError.deleteEntryFailed
        }

        await loadEntries()
    }

    func searchEntries(query: String) async -> [EntryModel] {
        guard let bridge = vaultBridge else { return [] }

        let summaries = await bridge.searchEntries(query: query)
        return summaries.map { summary in
            EntryModel(
                id: summary.id,
                title: summary.title,
                username: summary.username,
                favorite: summary.favorite
            )
        }
    }

    // MARK: - TOTP

    func generateTotp(entryId: String) async throws -> TotpCode {
        guard let bridge = vaultBridge else {
            throw VaultError.vaultLocked
        }

        guard let totp = await bridge.generateTotp(entryId: entryId) else {
            throw VaultError.totpFailed
        }

        return totp
    }

    // MARK: - Password Generation

    func generatePassword(length: Int, includeSymbols: Bool) async -> String? {
        return await VaultBridge.generatePassword(length: length, includeSymbols: includeSymbols)
    }

    func checkPasswordStrength(password: String) async -> PasswordAnalysis? {
        return await VaultBridge.checkPasswordStrength(password: password)
    }

    // MARK: - Biometric

    func enableBiometric() async throws {
        guard let bridge = vaultBridge else {
            throw VaultError.vaultLocked
        }

        // Generate biometric key
        let keyData = generateBiometricKey()
        let success = await bridge.setBiometricKey(keyData: keyData)

        guard success else {
            throw VaultError.biometricFailed
        }
    }

    func disableBiometric() async throws {
        guard let bridge = vaultBridge else {
            throw VaultError.vaultLocked
        }

        let success = await bridge.removeBiometricKey()
        guard success else {
            throw VaultError.biometricFailed
        }
    }

    func hasBiometricKey() async -> Bool {
        guard let bridge = vaultBridge else { return false }
        return await bridge.hasBiometricKey()
    }

    func unlockWithBiometric() async throws {
        guard let bridge = vaultBridge else {
            throw VaultError.vaultLocked
        }

        let success = await bridge.unlockWithBiometric()
        guard success else {
            throw VaultError.biometricFailed
        }

        isUnlocked = true
        await loadEntries()
    }

    // MARK: - Helpers

    private func generateBiometricKey() -> Data {
        // Generate 32 random bytes for biometric key
        var bytes = [UInt8](repeating: 0, count: 32)
        _ = SecRandomCopyBytes(kSecRandomDefault, 32, &bytes)
        return Data(bytes)
    }
}

// MARK: - Errors

enum VaultError: LocalizedError {
    case creationFailed
    case invalidPassword
    case vaultLocked
    case entryNotFound
    case addEntryFailed
    case updateEntryFailed
    case deleteEntryFailed
    case totpFailed
    case biometricFailed

    var errorDescription: String? {
        switch self {
        case .creationFailed:
            return "Failed to create vault"
        case .invalidPassword:
            return "Invalid master password"
        case .vaultLocked:
            return "Vault is locked"
        case .entryNotFound:
            return "Entry not found"
        case .addEntryFailed:
            return "Failed to add entry"
        case .updateEntryFailed:
            return "Failed to update entry"
        case .deleteEntryFailed:
            return "Failed to delete entry"
        case .totpFailed:
            return "TOTP generation failed"
        case .biometricFailed:
            return "Biometric operation failed"
        }
    }
}
